analyzers/src/SonarAnalyzer.Common/Rules/CommentsShouldNotBeEmptyBase.cs

```diff
@@ -0,0 +1,126 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class CommentsShouldNotBeEmptyBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S4663";
+
+    protected abstract string GetCommentText(SyntaxTrivia trivia);
+    protected abstract bool IsValidTriviaType(SyntaxTrivia trivia);
+
+    protected abstract bool IsSimpleComment(SyntaxTrivia trivia);
+    protected abstract bool IsEndOfLine(SyntaxTrivia trivia);
+    protected abstract bool IsWhitespace(SyntaxTrivia trivia);
+
+    protected override string MessageFormat => "Remove this empty comment";
+
+    protected CommentsShouldNotBeEmptyBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterTreeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            foreach (var token in c.Tree.GetRoot().DescendantTokens())
+            {
+                CheckTrivia(c, token.LeadingTrivia);
+                CheckTrivia(c, token.TrailingTrivia);
+            }
+        });
+
+    protected void CheckTrivia(SonarSyntaxTreeReportingContext context, IEnumerable<SyntaxTrivia> trivia)
+    {
+        foreach (var partition in Partition(trivia).Where(ShouldReport))
+        {
+            var start = partition.First().GetLocation().SourceSpan.Start;
+            var end = partition.Last().GetLocation().SourceSpan.End;
+
+            var location = Location.Create(context.Tree, TextSpan.FromBounds(start, end));
+            context.ReportIssue(Diagnostic.Create(Rule, location));
+        }
+    }
+
+    protected IEnumerable<IEnumerable<SyntaxTrivia>> Partition(IEnumerable<SyntaxTrivia> trivia)
+    {
+        var res = new List<List<SyntaxTrivia>>();
+
+        var current = new List<SyntaxTrivia>();
+        var endOfLineFound = false;
+
+        foreach (var trivium in trivia)
+        {
+            if (IsWhitespace(trivium))
```

Comment:
Isn't it better to skip if:
```suggestion
            if (!(IsAnyComment(trivia) || IsEndOfLine(trivia)))
```

Reply:
what about this case?
```csharp
// <- simple comment trivia
region #MY_REGION //<- prepro-directive trivia
// <- simple comment trivia
```
I do not want to skip in the case of non-comment and non-EOL trivia, I want to go to the last else block and reset my current partition.

Reply:
Makes sense.